pytorch_lightning/callbacks/batch_size_finder.py

```diff
@@ -0,0 +1,423 @@
+        self.scale_batch_size(trainer, pl_module)
+
+    def on_predict_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
+        self.scale_batch_size(trainer, pl_module)
+
+    def _adjust_batch_size(
+        self,
+        trainer: "pl.Trainer",
+        factor: float = 1.0,
+        value: Optional[int] = None,
+        desc: Optional[str] = None,
+    ) -> Tuple[int, bool]:
+        """Helper function for adjusting the batch size.
+
+        Args:
+            trainer: instance of pytorch_lightning.Trainer
+            factor: value which the old batch size is multiplied by to get the
+                new batch size
+            value: if a value is given, will override the batch size with this value.
+                Note that the value of `factor` will not have an effect in this case
+            desc: either ``"succeeded"`` or ``"failed"``. Used purely for logging
+
+        Returns:
+            The new batch size for the next trial and a bool that signals whether the
+            new value is different than the previous batch size.
+        """
+        from pytorch_lightning.trainer.states import TrainerFn
+
+        model = trainer.lightning_module
+        batch_size = lightning_getattr(model, self.batch_arg_name)
+        new_size = value if value is not None else int(batch_size * factor)
+        if desc:
+            rank_zero_info(f"Batch size {batch_size} {desc}, trying batch size {new_size}")
+
+        # TODO improve this for eval CombinedLoader and multi dataloaders
+        if trainer.state.fn == TrainerFn.FITTING:
+            if not self._is_valid_batch_size(new_size, trainer.train_dataloader, trainer):
+                new_size = min(new_size, len(trainer.train_dataloader.dataset))
+        if trainer.state.fn == TrainerFn.VALIDATING:
+            if not self._is_valid_batch_size(new_size, trainer.val_dataloaders[0], trainer):
+                new_size = min(new_size, len(trainer.val_dataloaders[0].dataset))
+        if trainer.state.fn == TrainerFn.TESTING:
+            if not self._is_valid_batch_size(new_size, trainer.test_dataloaders[0], trainer):
+                new_size = min(new_size, len(trainer.test_dataloaders[0].dataset))
+        if trainer.state.fn == TrainerFn.PREDICTING:
+            if not self._is_valid_batch_size(new_size, trainer.predict_dataloaders[0], trainer):
+                new_size = min(new_size, len(trainer.predict_dataloaders[0].dataset))
+
+        changed = new_size != batch_size
+        lightning_setattr(model, self.batch_arg_name, new_size)
+        return new_size, changed
+
+    @staticmethod
+    def _is_valid_batch_size(batch_size: int, dataloader: DataLoader, trainer: "pl.Trainer") -> bool:
+        from pytorch_lightning.utilities.data import has_len_all_ranks
+
+        module = trainer.lightning_module or trainer.datamodule
+        return not has_len_all_ranks(dataloader, trainer.strategy, module) or batch_size <= len(dataloader)
```

Comment:
@rohitgr7 This shouldn't be enough as has_len_all_ranks validates the additional length isn't bigger than the dataset. We also need to check that each batch_size found for each rank is the same. Otherwise, the DistributedSampler on each rank won't compute the same chunks and data would be duplicated.

Reply:
will address this with DDP+batch size finder PR.